python_modules/libraries/dagster-dbt/dagster_dbt/asset_decorators.py

```diff
@@ -0,0 +1,86 @@
+from dataclasses import dataclass
+from functools import cached_property
+from typing import Callable, Mapping, Optional
+
+from dagster import AssetsDefinition, multi_asset
+from dagster._core.definitions.events import AssetKey
+
+from .asset_utils import default_asset_key_fn, get_dbt_multi_asset_args, get_deps
+from .utils import ASSET_RESOURCE_TYPES, output_name_fn, select_unique_ids_from_manifest
+
+
+@dataclass
+class DbtManifest:
+    """Helper class for dbt manifest operations."""
+
+    manifest: dict
+
+    @cached_property
+    def dbt_unique_id_to_nodes(self) -> Mapping[str, dict]:
+        """A mapping of a dbt node's unique id to the node's dictionary representation in the manifest.
+        """
+        return {
+            **self.manifest["nodes"],
+            **self.manifest["sources"],
+            **self.manifest["exposures"],
+            **self.manifest["metrics"],
+        }
+
+    @cached_property
+    def asset_key_to_nodes(self) -> Mapping[AssetKey, dict]:
+        """A mapping of the default asset key for a dbt node to the node's dictionary representation in the manifest.
+        """
+        return {default_asset_key_fn(node): node for node in self.dbt_unique_id_to_nodes.values()}
+
+    @cached_property
+    def output_name_to_nodes(self) -> Mapping[str, dict]:
+        """A mapping of the default output name for a dbt node to the node's dictionary representation in the manifest.
+        """
+        return {output_name_fn(node): node for node in self.dbt_unique_id_to_nodes.values()}
+
+    def get_node_by_output_name(self, output_name: str) -> dict:
+        """Get a dbt node's dictionary representation in the manifest by its Dagster output name."""
+        return self.output_name_to_nodes[output_name]
+
+    def get_node_by_asset_key(self, asset_key: AssetKey) -> dict:
+        """Get a dbt node's dictionary representation in the manifest by its Dagster output name."""
+        return self.asset_key_to_nodes[asset_key]
+
+
+def dbt_multi_asset(
+    *,
+    manifest: dict,
+    select: str = "*",
+    exclude: Optional[str] = None,
+) -> Callable[..., AssetsDefinition]:
+    unique_ids = select_unique_ids_from_manifest(
+        select=select, exclude=exclude or "", manifest_json=manifest
+    )
+    dbt_nodes = DbtManifest(manifest).dbt_unique_id_to_nodes
+
+    deps = get_deps(
+        dbt_nodes,
+        selected_unique_ids=unique_ids,
+        asset_resource_types=ASSET_RESOURCE_TYPES,
```

Comment:
I don't feel great about having this here, since we're making a decision for the user on what deserves to be an `AssetKey`. I think ideally, this is a choice that the user should make.

Reply:
yeah this is a good callout. my thinking is that the desired `asset_resource_types` are highly correlated to the commands that you're running inside the compute_fn.

this is how it works right now (if you're running `dbt build`, then asset resource types = the set of things that dbt build runs, if you're running `dbt build`, then asset resource types = just models).

I also think that the command you're running is highly correlated to what actually exists in your project. That is, if you have seeds and snapshots, you're much more likely to use `dbt build` than `dbt run`, or in this new world, potentially run all of `dbt run` ,  `dbt seed`, and `dbt snapshot`.

In the (theoretically somewhat rare) case that people do have (e.g.) snapshots, but don't want them to be loaded as Dagster assets, it's easy enough to have `exclude="resource:snapshot"`, so I think this would be an acceptable way to configure this behavior.

Reply:
the one potential exception here would be dbt tests, which are executed, but are not represented as assets. In theory, someone might want to alter that behavior, and modifying select/exclude would not help them there. I don't think we need to worry about this in the short term though.